### PR TITLE
Issue #339. with tn.DefaultBackend(backend): support

### PR DESCRIPTION
**Pull request by katolikyan**
I tried to implement this one. I believe the DefaultBackend class could be better but still :)

I also have tried to implement it in another way, more like a temporary change for `config.default_backend` in a `with` statement. In this way, it doesn't affect any code inside the `class Node`, but if someone will set `config.default_backend` inside a `with` statement that will cause a little mess. Here is the main idea how would another approach look like:

```python3
from typing import Text, Union
from contextlib import contextmanager
from tensornetwork.backends.base_backend import BaseBackend
import tensornetwork.config as config
from tensornetwork.__init__ import set_default_backend

@contextmanager
def DefaultBackend(backend: Union[Text, BaseBackend]) -> None:
    try:
        default_backend = config.default_backend
        set_default_backend(backend)
        yield None
    finally:
        set_default_backend(default_backend)
```

### CODE REVIEW

**Comment by katolikyan** [comment]
@googlebot I signed it!

**Comment by katolikyan** [comment]
@googlebot I fixed it.

File: tensornetwork/backend_contextmanager.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+    def __init__(self, backend: Union[Text, BaseBackend]) -> None:
+        DefaultBackend.backend = backend
+
+    def __enter__(self):
+        DefaultBackend.with_statement = True
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        DefaultBackend.with_statement = False
+        DefaultBackend.backend = None
```
Comment (Thenerdstation) [inline tensornetwork/backend_contextmanager.py]: Nice! I like the object oriented approach.

However, I think we should be using a stack based approach for finding the current backend. (For an example, see what we did with the `NodeCollection` [code](https://github.com/google/TensorNetwork/blob/83ff4af11dcd03d6c0db70e139c7137154dc015d/tensornetwork/network_components.py#L1978).

File: tensornetwork/tests/backend_contextmanager_test.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    with tn.DefaultBackend("numpy"):
+
+        assert tn.config.default_backend == "pytorch"
+
+        a = tn.Node(np.ones((10,)))
+        tn.set_default_backend("tensorflow")
+        b = tn.Node(np.ones((10,)))
+
+    assert (type(a.backend) == type(b.backend)) == True
+    assert tn.config.default_backend == "tensorflow"
```
Comment (Thenerdstation) [inline tensornetwork/tests/backend_contextmanager_test.py]: We should also consider the case where we have nested `with` statements. You can imagine this happening for when a library needs to implement something purely in numpy or when a user wants to test their code in all backends.

Maybe add a test like this?

```python
with tn.DefaultBackend("tensorflow"):
  a = tn.Node(...)
  assert a.backend.name == "tensorflow" 
  with tn.DefaultBackend("numpy"):
     b = tn.Node(...)
     assert b.backend.name == "numpy"
  c = tn.Node(...)
  assert c.backend.name == "tensorflow"
```

**Comment by katolikyan** [comment]
Thank you for the comments! I tried to implement it with a stack as you showed in the example.
I also have added a couple of tests.
But I am not sure if I should have written the `DefaultBackend` class inside the `network_components.py` and put the stack into the `ops.py` file.

File: tensornetwork/tests/backend_contextmanager_test.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  assert final_node.tensor == 10.0
+
+
+def test_contextmanager_interaption():
+  tn.set_default_backend("pytorch")
+  assert tn.config.default_backend == "pytorch"
+  with tn.DefaultBackend("numpy"):
+    assert tn.config.default_backend == "pytorch"
+    a = tn.Node(np.ones((10,)))
+    tn.set_default_backend("tensorflow")
```
Comment (Thenerdstation) [inline tensornetwork/tests/backend_contextmanager_test.py]: using `set_default_backend` while there is a backend stack feels like a foot gun. Can you raise an error in this case instead?

File: tensornetwork/network_components.py [hunk 4]
```diff
@@ -516,7 +517,11 @@ def __init__(self,
       backend = tensor.backend
       tensor = tensor.tensor
     if not backend:
-      backend = config.default_backend
+      default_backend = _default_backend_stack.get_current_backend()
+      if default_backend:
+        backend = default_backend.backend
+      else:
+        backend = config.default_backend
```
Comment (Thenerdstation) [inline tensornetwork/network_components.py]: Can this logic be moved to the `DefaultBackend` code?

File: tensornetwork/backend_contextmanager.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    _default_backend_stack.stack.pop()
+
+class _DefaultBackendStack():
+  """A stack to keep track default backends context manager"""
+
+  def __init__(self):
+    self.stack = []
+
+  def get_current_backend(self):
+    return self.stack[-1] if self.stack else None
```
Comment (Thenerdstation) [inline tensornetwork/backend_contextmanager.py]: Move the logic in network_components to here.

File: tensornetwork/tests/backend_contextmanager_test.py (hunk 3 above)
Comment (katolikyan) [reply tensornetwork/tests/backend_contextmanager_test.py]: Sorry I didn't get what you mean by raising an error in this case. Prohibit using `set_default_backend` while in `with` statement?

**Comment by katolikyan** [comment]
I moved the logic to the right place. and cleaned up a test a little to get rid of some unnecessary checks.

File: tensornetwork/tests/backend_contextmanager_test.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      assert b.backend.name == "numpy"
+    c = tn.Node(np.ones((10,)))
+    assert c.backend.name == "tensorflow"
+  d = tn.Node(np.ones((10,)))
+  assert d.backend.name == "numpy"
+
+def test_contextmanager_wrong_item():
+  a = tn.Node(np.ones((10,)))
+  try:
+    with tn.DefaultBackend(a):
```
Comment (Thenerdstation) [inline tensornetwork/tests/backend_contextmanager_test.py]: Add `# pytype: disable=wrong-arg-types` to this line to pass the build.

File: tensornetwork/tests/backend_contextmanager_test.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      pass
+    assert False
+  except ValueError:
+    assert True
```
Comment (Thenerdstation) [inline tensornetwork/tests/backend_contextmanager_test.py]: You should use `with pytest.raises(ValueError)` instead.

https://docs.pytest.org/en/latest/assert.html

File: tensornetwork/backend_contextmanager.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
+from typing import Text, Union
+from tensornetwork.backends.base_backend import BaseBackend
+import tensornetwork.config as config
```
Comment (Thenerdstation) [inline tensornetwork/backend_contextmanager.py]: Can we move the `set_default_backend` and everything in `config` to this file? No need to make these separate anymore and I like this filename more.

File: tensornetwork/__init__.py [hunk 9]
```diff
@@ -15,6 +15,7 @@
 from tensornetwork.utils import load_nodes, save_nodes
 from tensornetwork.matrixproductstates.finite_mps import FiniteMPS
 from tensornetwork.matrixproductstates.infinite_mps import InfiniteMPS
+from tensornetwork.backend_contextmanager import DefaultBackend
 
 
 def set_default_backend(backend: Union[Text, BaseBackend]) -> None:
```
Comment (Thenerdstation) [inline tensornetwork/__init__.py]: Add a check that if the default backend stack is empty, and if it's not, raise an `AssertionError`. 

This is to help avoid the confusing situation like: 

```python
with tn.DefaultBackend("tensorflow"):
  tn.set_default_backend("jax")
  a = tn.Node(...)
```
Should `a` be using jax or tensorflow? It's confusing and kinda a footgun, so let's just crash with a clear message instead.

**Comment by katolikyan** [comment]
Hopefully, I haven't messed up the code. Removed `config` imports with the thought that it became useless in this case. But feel like maybe I shouldn't have done that.

Also, from my understanding, we don't need direct access to `default_backend` anymore, do we?
I moved it to the `_DefaultBackendStack` class, trying to avoid using `global`. I can quickly rewrite it as a global with something like this if needed :)

```python3
default_backend = "numpy"

def set_default_backend(backend: Union[Text, BaseBackend]) -> None:
  global default_backend
  if _default_backend_stack.stack:
    raise AssertionError("The default backend should not be changed "
                         "inside the backend context manager")
  if not isinstance(backend, (Text, BaseBackend)):
    raise ValueError("Item passed to set_default_backend "
                     "must be Text or BaseBackend")
  default_backend = backend
```

**Comment by Thenerdstation** [comment]
Looks awesome! Thank you so much! I've been waiting on this one for a while. ^-^

**Comment by katolikyan** [comment]
Glad to hear this :) it was nice to work on that issue, thank you! :)

**Pull request merged**